### ISSUE TITLE
KeyError in Element.remove()

### ISSUE THREAD

**Issue by mjchl** [zauberzeug/nicegui#2814]
### Description

The following error occurred in my application with the new log implementation in v.1.4.19.
```
    log.push(message)
  File "/usr/local/lib/python3.12/site-packages/nicegui/elements/log.py", line 29, in push
    self.remove(0)
  File "/usr/local/lib/python3.12/site-packages/nicegui/element.py", line 526, in remove
    self.client.remove_elements(elements)
  File "/usr/local/lib/python3.12/site-packages/nicegui/client.py", line 302, in remove_elements
    del self.elements[element_id]
        ~~~~~~~~~~~~~^^^^^^^^^^^^
KeyError: 35
```

I wasn't able to reproduce it on purpose but apparently it is possible under some circumstances that `Element._collect_descendants()` returns Elements that are not (any longer) linked to the client. I don't know how the deletion of Elements is done and maybe it is sufficient to make sure that only Elements not marked deleted are returned by `Element._collect_descendants()`. In any case if the removal of any child element fails for whatever reason, the element is not removed from its parent_slot, but due to the order of removal in `Client.remove_elements()` it is actually already deleted from the client itself.

**Comment by falkoschindler** [comment]
Thanks for reporting this issue, @mjchl!

It's strange though, because `ui.log` simply calls `self.remove(0)` to remove the first element - nothing special.
Do you, by any chance, use multi-threading? Manipulating the UI from multiple threads could indeed lead to unexpected behavior.

In any case, I'd like to find a logical explanation first before changing the implementation.

**Comment by mjchl** [comment]
I don't use multi-threading, but I do try to share the entries between multiple page instances with their own logs.
A minimal example would look something like this:
```python
from nicegui import ui, app
from datetime import datetime as dt
from collections import deque


MAX_ENTRIES = 1024

class SharedLog:
    def __init__(self) -> None:
        self.instances = {}
        self.entries = deque(maxlen=MAX_ENTRIES)
    def register(self, id, log):
        self.instances[id] = log
    def unregister(self, id):
        if id in self.instances:
            self.instances.pop(id)
    def push(self, msg: str):
        self.entries.append(msg)
        for log in self.instances.values():
            log.push(msg)

@ui.page('/')
def page(client):
    log = ui.log(max_lines=MAX_ENTRIES)
    for entry in shared_log.entries:
        log.push(entry)
    shared_log.register(client.id, log)

shared_log = SharedLog()
app.on_disconnect(lambda c: shared_log.unregister(c.id))
ui.timer(0.5, lambda: shared_log.push(f"{dt.now():%X} -> new entry"))

ui.run()
```
Now that I think about it again I would assume that the problem occurs when a client disconnects and a log push occurs before `app.on_disconnect()` callback is executed.

**Comment by mjchl** [comment]
Yes I'm also not able to reproduce it on purpose. What I know is that there where multiple clients connected when it happened and it then continued to occur on every log entry from that point until I stopped it. Which is also odd since my possible dangling references should not be used anymore if a disconnect was actually the cause.

 I will try to add more debug information if it occurs again.

**Comment by mjchl** [comment]
I've noticed this type of error now appearing in my application:
```
'084e3270-6398-4dc1-a88c-fbf7b18aa2c3'
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/nicegui/background_tasks.py", line 52, in _handle_task_result
    task.result()
  File "/usr/local/lib/python3.12/site-packages/nicegui/client.py", line 257, in handle_disconnect
    self.delete()
  File "/usr/local/lib/python3.12/site-packages/nicegui/client.py", line 316, in delete
    del Client.instances[self.id]
        ~~~~~~~~~~~~~~~~^^^^^^^^^
KeyError: '084e3270-6398-4dc1-a88c-fbf7b18aa2c3'
```
I have a counter to keep track of the number of registrations to the shared log. This counter increases (not always though) with the above error. That means that the page function is executed for a new client, but there is never a disconnect event. This should be the source of my dead references. But I have no idea what causes this error.

There weren't any "new" connections happening, only one was kept open so my best guess would be that there is some kind of reconnection happening where a new client instance is created but then the connection is mapped to the existing client object.

**Comment by mjchl** [comment]
Hello again. I could reproduce wild connection behaviour by introducing some blocking function calls to the system.
I used following code:
```python
from nicegui import ui
from datetime import datetime as dt
from collections import deque
import time
import asyncio


MAX_ENTRIES = 128

class SharedLog:
    def __init__(self) -> None:
        self.pages = {}
        self.entries = deque(maxlen=MAX_ENTRIES)
        self._running = 0
        self._c = 0

    @property
    def work(self):
        return f"{self._running} workers active"

    def register(self, cid, log):
        self.pages[cid] = log
        self._c += 1

    def unregister(self, cid):
        if cid in self.pages:
            self.pages.pop(cid)
            self._c -= 1

    def push(self, msg: str):
        msg = f"{msg}: {self._c}"
        self.entries.append(msg)
        for log in self.pages.values():
            log.push(msg)

    def some_blocking(self):
        self._running += 1
        async def _blocking():
            while True:
                time.sleep(1)
                await asyncio.sleep(5)
        t = asyncio.create_task(_blocking())
        t.add_done_callback(self._done)

    def _done(self, result):
        print(result)

@ui.page('/')
def page(client):
    with ui.row():
        ui.button("more work!", on_click=shared_log.some_blocking)
        ui.label().bind_text_from(shared_log, 'work')
    log = ui.log(max_lines=MAX_ENTRIES)
    for entry in shared_log.entries:
        log.push(entry)
    client.on_connect(lambda c: shared_log.register(c.id, log))
    client.on_disconnect(lambda c: shared_log.unregister(c.id))

shared_log = SharedLog()
ui.timer(1, lambda: shared_log.push(f"{dt.now():%X}"))

ui.run()
```
The log writes the number of subscriptions to the shared log. For me adding 3 occasionally blocking loops with the button is enough to make the number of subscriptions growing. I assume that the real client makes several reconnect attempts and then multiple pending handshakes are done with the same client. While blocking functions can be avoided, there might be times when a system is unresponsive from external reasons.

The `KeyError` on client deletion is caused by multiple calls to the disconnect handler, which to my understanding does not check if there is already a disconnect task scheduled.

**Comment by falkoschindler** [comment]
@mjchl If I understand correctly, your code adds more and more blocking code (`time.sleep(1)`) to the main thread. This is usually a bad idea, as discussed in [this FAQ](https://github.com/zauberzeug/nicegui/wiki/FAQs#how-to-not-get-the-reconnect-popup-or-avoid-laggy-interaction).

Anyway, I don't get any `KeyError`, just the "Connection lost." popup. But if removing blocking code from the main thread solves your problem, we can certainly close this issue - like you just did. 👍🏻

**Comment by mjchl** [comment]
@falkoschindler the code also logs the number of subscriptions to the shared log. It increses on all clients `on_connect` and decreases on all `on_diconnect` callbacks. When you increase the amount of blocking code, the number of subscriptions starts to increase without real new connections.

**Comment by falkoschindler** [comment]
I just found a reproduction for the `KeyError` when pushing to `ui.log`:

```py
from typing import Callable
from nicegui import ui

callbacks: list[Callable] = []

@ui.page('/page')
def page():
    log = ui.log(3)
    callbacks.append(lambda: log.push('X'))

ui.button('Log', on_click=lambda: [callback() for callback in callbacks])

ui.run()
```

1. Visit `/`.
2. Visit `/page` in a new tab.
3. Click "Log" multiple times. Up to three "X"s appear in the log.
4. Close the `/page` tab. Wait 3 seconds for the client to disconnect.
5. Click "Log" once again. The callback tries to push to the log element on a disconnected client, which has already been emptied during disconnect. This causes a KeyError.

Long story short: When a KeyError in `Client.remove_elements` occurs, there is probably some kind of memory leak holding references to old client instances.
